### ISSUE TITLE
Use backpressure_detection mechanism to deal with fast publishing messages. It might be a bug of pika

### ISSUE THREAD

**Issue by yangyuliang1981** [pika/pika#728]
Use backpressure_detection mechanism to deal with fast publishing messages.
I guess it is a bug of pika
1. I used add_backpressure_callback,but it raised an exception.
2. I added callback functions **add_on_connection_blocked_callback *_and *_add_on_connection_unblocked_callback**.
   But they are not triggered.
   
   In pika docs, it said "Publishers can use this to voluntarily suspend publishing, instead of relying on back pressure throttling. The callback will be passed the Connection.Blocked method frame."
   
   I think I used them in wrong ways. Any suggestions??

```
import logging
import pika
import time

logging.basicConfig(level=logging.DEBUG)


def on_open(new_connection):
    global connection
    connection = new_connection
    connection.channel(on_channel_open)
    connection.add_backpressure_callback(sleep_seconds)
    connection.add_on_connection_blocked_callback(conn_lock)
    connection.add_on_connection_unblocked_callback(conn_unlock)
    connection.add_timeout(1, publish)


def on_channel_open(new_channel):
    global channel
    channel = new_channel
    channel.exchange_declare(exchange='exchange_name', exchange_type='topic')


def sleep_seconds():
    time.sleep(1)


def conn_lock():
    print 'lock'


def conn_unlock():
    print 'unlock'


def publish():
    global connection
    global channel

    channel.basic_publish(exchange='exchange_name',
                          routing_key='routing_key',
                          body='Test Message' * 100)
    print 'send'
    connection.add_timeout(0.001, publish)


channel = None
parameters = pika.ConnectionParameters(host='192.168.91.71', backpressure_detection=True)
connection = pika.SelectConnection(parameters=parameters, on_open_callback=on_open)
connection.ioloop.start()

```

**Exceptions:**
`
File "C:\Python27\lib\site-packages\pika-0.10.0-py2.7.egg\pika\connection.py", line 1062, in _detect_backpressure
    int(buffer_size / avg_frame_size))
ZeroDivisionError: integer division or modulo by zero
`

**Comment by yangyuliang1981** [comment]
Using the latest code from github, it works ok, while the code from pypi is not work.
The former fix some bugs with connection close.

Not test backpressure yet, another issue.

**Comment by vitaly-krugl** [comment]
@yangyuliang1981, backpressure will likely get deprecated. The blocked/unblocked notifications are the preferred mechanism.

**Comment by vitaly-krugl** [comment]
RabbitMQ's Connection.Blocked is different from TCP/IP stream back-pressure.

RabbitMQ sends Connection.Blocked to publishers when it is running low on resources (e.g., low memory or disk space). If you registered via `add_on_connection_blocked_callback` before Connection.Blocked arrived, then your callback should be called upon arrival of this message. When a connection is blocked via Connection.Blocked, RabbitMQ won't process any incoming traffic from that connection (commands/messages/acks/nacks/etc., nothing) on that connection until the connection becomes unblocked. When the connection becomes unblocked, RabbitMQ will send Connection.Unblocked.

 RabbitMQ uses back-pressure when it can't keep up with incoming data on a connection (simply stops reading from connection until it catches up).

**Comment by vitaly-krugl** [comment]
@gmr
Hi Gavin, we discussed recently the possibility of deprecating the backpressure detection mechanism in pika, and I added deprecation comments as the result of that discussion. However, some (new?) mechanism may still be needed to help users of async adapters avoid out-of-memory conditions resulting from buffering of outbound data in the connection.

Imagine an app that reads data from TCP/IP stream and publishes it to AMQP, and assume that the source data is coming in much faster than RabbitMQ is able to absorb. In this case, RabbitMQ will be applying backpressure, and the pika connection's outbound buffer will be growing continuously. This will lead to out-of-memory condition in the app eventually.

Perhaps a simple, user-configurable high-water/low-water notification mechanism may be created for outbound buffer that replaces the current somewhat more complicated backpressure detection mechanism?

**Comment by vitaly-krugl** [comment]
@yangyuliang1981, would you mind adding a unit test to tests/unit/connection_tests.py that reproduces the ZeroDivisionError that you reported?

**Comment by vitaly-krugl** [comment]
There is clearly a BUG in `Connection. _send_message` in pika maser that is likely responsible for this `ZeroDivisionError` exception.  `Connection. _send_message` updates `self.frames_sent`, but not `self.bytes_sent`. Compare with `Connection._send_frame`.

**Comment by vitaly-krugl** [comment]
@gmr, this bug (see my previous comment) needs to be fixed before next release.

**Comment by yangyuliang1981** [comment]
@vitaly-krugl, It's ok to add the case as an unit test.  
Should I add it, or you will do it??

I use the latest code from github, then the connection is not closed strangely. But it takes 5 seconds to send and recv a message. It introduce another bug obviously.

**Comment by vitaly-krugl** [comment]
@yangyuliang1981, the 5 seconds to send/receive a message sounds like something may be wrong with your RabbitMQ or network configuration. Pika runs many acceptance tests that send/receive many messages. If they took 5 seconds per message, then the acceptance test runtimes would be very long. Here is an example set of recent test runs: https://travis-ci.org/pika/pika/builds/115000880 for your reference.

To test pika's performance on Travis, implement a simple send/receive stress test with timing in pika/tests/acceptance/ directory and submit a dummy PR. Travis instances are pretty slow, but you shouldn't get anything close to 5 seconds per message.

**Comment by vitaly-krugl** [comment]
@yangyuliang1981, the fix should be very trivial

First, implement a couple of simple unit tests in tests/unit/connection_tests.py (mocking out the actual I/O like the other tests in that file) :
1. Test that Connection._send_message updates both self.frames_sent and self.bytes_sent correctly and run the test to confirm that it detects the inconsistency in self.bytes_sent.
2. Test `Connection._detect_backpressure` after calling `Connection._send_message`. Run the test against current code base to confirm that it fails with `ZeroDivisionError`.

Then, implement the simple fix in `Connection._send_message`.  Right after `self.frames_sent += len(write_buffer)`, add:

```
self._bytes_sent += sum(len(buff) for buff in write_buffer)
```

**Comment by yangyuliang1981** [comment]
@vitaly-krugl thanks, you explained it very clearly. But I don't familiy with the mock things and I don't know how to submit a PR. Can you help me to do these things?

I created the **5 seconds issue**, and you have closed it as pika not support thread.
Then I am confused about how to schedule the publishing process.

The following code is a segment of a pika publishing example.
Does it means I should publish the message with following mode?
First, create a global message queue,
Second, start a loop to checking the queue until it is not empty,  and add a timeout callback function
Third, run the basic_publish function to send the message
Forth, go back to second step for the checking loop.

Is the above mode correct?
I am worried about the message queue checking loop will take too much CPU resources.

   ```
def schedule_next_message(self):
        """If we are not closing our connection to RabbitMQ, schedule another
        message to be delivered in PUBLISH_INTERVAL seconds.

        """
        if self._stopping:
            return
        LOGGER.info('Scheduling next message for %0.1f seconds',
                    self.PUBLISH_INTERVAL)
        self._connection.add_timeout(self.PUBLISH_INTERVAL,
                                     self.publish_message)

def publish_message(self):       
        if self._stopping:
            return

        message = {u'مفتاح': u' قيمة',
                   u'键': u'值',
                   u'キー': u'値'}
        properties = pika.BasicProperties(app_id='example-publisher',
                                          content_type='application/json',
                                          headers=message)

        self._channel.basic_publish(self.EXCHANGE, self.ROUTING_KEY,
                                    json.dumps(message, ensure_ascii=False),
                                    properties)

        self._message_number += 1
        self._deliveries.append(self._message_number)
        LOGGER.info('Published message # %i', self._message_number)
        self.schedule_next_message()
```

**Comment by vitaly-krugl** [comment]
@yangyuliang1981, mock is documented here http://www.voidspace.org.uk/python/mock/.

Regarding
> Then I am confused about how to schedule the publishing process

the answer depends on where the data that you want to publish comes from and the programming model that your app is using. SelectConnection, TornadoConnection, and TwistedConnection adapters are built around the concept of "ioloop" event loop mechanism, where the application uses the primitives provided by the specific ioloop to schedule and perform all I/O and other operations in the same thread.

For example, if your data comes from a TCP/IP socket, you would create and associate a pika async adapter (one of Select/Tornado/TwistedConnection), and then start the ioloop (e.g., `ioloop.start()`), which blocks inside the ioloop's using `select`, epoll, kqueue, or similar mechanism, waiting for registered events of interest to happen. When the AMQP connection is established, your app would create a pika channel, and when the channel creation completes, your app would create/connect a non-blocking input socket, then register the input socket with the ioloop to notify your app when the socket is readable. When your input socket becomes readable, the ioloop notifies your code (e.x., via callback function), your code reads data from the socket, parses it into messages, and publishes it via the async pika connection's channel. The idea behind the event loop concept is that it reduces the amount of context switches and memory used by applications versus the multi-threaded model. Code that works best with this concept is I/O-bound code that uses the ioloop for all blocking operations, such as waiting for sockets to become readable, writeable, etc.

Another model is the blocking model, which might be more appropriate if your data comes from a source that doesn't work with ioloop concept, or you don't need the performance benefit of the ioloop and wish to simplify your logic instead. For example, let's say your code reads data from MySQL and publishes to AMQP. Pika has an adapter for this called `pika.BlockingConnection`. See https://www.rabbitmq.com/tutorials/tutorial-three-python.html and http://pika.readthedocs.org/en/0.10.0/examples.html.

**Comment by vitaly-krugl** [comment]
Hi @yangyuliang1981, in the interest of time, I fixed this via PR #738. Please review. Thank you.

**Comment by yangyuliang1981** [comment]
ok, I'll run it later!

And I think you should add your explanation about how to use pika into the official docs.
Then the fresh guys, like me, can learn how to use pika quickly.

**Comment by yangyuliang1981** [comment]
@vitaly-krugl 
As you said, "your app would create/connect a non-blocking input socket, then register the input socket with the ioloop to notify your app when the socket is readable. When your input socket becomes readable, the ioloop notifies your code (e.x., via callback function), your code reads data from the socket, parses it into messages, and publishes it via the async pika connection's channel, and returns control back to the ioloop."

I try to realize it ,but not works.
The code in selectconnection just connect 
The error is "OSError: [Errno 9] Bad file descriptor"
```
# in open channel callback
read_sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
read_sock.setblocking(0)
read_sock.connect(('127.0.0.1', 0))
connection.ioloop.add_handler(DeviceIO.read_sock.fileno(), DeviceIO.read_interrupt2, DeviceIO.connection.READ)

# in read_interrunpt2 function
os.read(interrupt_sock, 512)

# in another thread, try to send data
os.write(read_sock.fileno(), 'hello')
```

**Comment by vitaly-krugl** [comment]
@yangyuliang1981 Regarding
> @vitaly-krugl 
> As you said, "your app would create/connect a non-blocking input socket, then register the input socket with the ioloop to notify your app when the socket is readable. When your input socket becomes readable, the ioloop notifies your code (e.x., via callback function), your code reads data from the socket, parses it into messages, and publishes it via the async pika connection's channel, and returns control back to the ioloop."
>
>I try to realize it ,but not works.
The code in selectconnection just connect 
The error is "OSError: [Errno 9] Bad file descriptor"
>
>
```
# in open channel callback
read_sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
read_sock.setblocking(0)
read_sock.connect(('127.0.0.1', 0))
connection.ioloop.add_handler(DeviceIO.read_sock.fileno(), DeviceIO.read_interrupt2, DeviceIO.connection.READ)
>
# in read_interrunpt2 function
os.read(interrupt_sock, 512)
>
# in another thread, try to send data
os.write(read_sock.fileno(), 'hello')
```

@yangyuliang1981, your datagram programming code isn't quite right. With datagram socket communication (unlike TCP streams), there is no real connection made. You can use connect on a datagram socket to set up the destination address, so that you can use send instead of sendto. The receiving datagram socket needs to be bound on that address and port number via bind. See Sending and Receiving here: https://wiki.python.org/moin/UdpCommunication

**Comment by vitaly-krugl** [comment]
@yangyuliang1981, also note that datagram socket communication is not reliable; packets may get lost and there are no retries.

**Comment by yangyuliang1981** [comment]
@vitaly-krugl 
Yes, when cpu is busy, and in some other situations, the udp package will be lost.
And It means I should change the pika code in several functions too.

UDP works well with pika.
And I  tried to use TCP to do the work, but not works by now. I will try to figure it out later.

I test **backpressure_detection **issue again, it works well by now.
Thanks for your help.

**Comment by vitaly-krugl** [comment]
Thank you, @yangyuliang1981

**Comment by vitaly-krugl** [comment]
I create pull request #1042 to clarify the semantics of add_on_connection_blocked_callback and add_on_connection_unblocked_callback.

**Comment by vitaly-krugl** [comment]
Backpressure detection has been fixed since the opening of this issue, so closing it. If the issue still exists in the latest release (0.12.0), please reopen.
